File: frontend/src/types/blueprint.ts
export interface BlueprintConfig {
  architectureStyle: string;
  database: string;
  backend: string;
  frontend: string;
  cloudProvider: string;
  projectSize: string;
}

export interface Column {
  name: string;
  type: string;
  primaryKey: boolean;
  foreignKey: string | null;
  unique: boolean;
  nullable: boolean;
  notes?: string;
}

export interface Table {
  name: string;
  columns: Column[];
}

export interface ApiEndpoint {
  method: string;
  route: string;
  description: string;
  authRequired: boolean;
  validation: string;
  headers: Array<{ name: string; description: string }>;
  statusCodes: Array<{ code: number; description: string }>;
  sampleRequest: string;
  sampleResponse: string;
  errors: Array<{ code: number; message: string }>;
}

export interface TimelinePhase {
  phase: string;
  description: string;
  days: number;
}

export interface PromptAnalysis {
  projectName: string;
  industry: string;
  businessType: string;
  complexity: string;
  expectedUsers: string;
  scale: string;
  budget: string;
  cloudRequirements: string;
  compliance: string;
  estimatedTimeline: string;
}

export interface Decision {
  component: string;
  userRequirement: string;
  recommendation: string;
  alternatives: string[];
  reason: string;
}

export interface PrdStructure {
  documentMetadata: {
    ownership: string;
    deploymentTarget: string;
    versionStatus: string;
  };
  executiveSummary: string;
  userStories: Array<{ persona: string; story: string }>;
  businessRules: Array<{ rule: string }>;
  acceptanceCriteria: Array<{ feature: string; criteria: string[] }>;
  uxDesign: {
    interfaceOverview: string;
    layoutDescription: string;
  };
  businessFlow: string[];
  systemFlow: string[];
}

export interface Monitoring {
  tracing: string;
  metrics: string[];
  dashboards: string[];
  healthChecks: string[];
}

export interface EstimatedCost {
  aws: string;
  development: string;
  duration: string;
}

export interface AiRecommendations {
  alternativeTechStack: string[];
  potentialBottlenecks: string[];
  scalingAdvice: string[];
  securityAdvice: string[];
  estimatedComplexity: string;
  architectureScore: string;
}

export interface Mermaid {
  erDiagram: string;
  architectureDiagram: string;
  flowDiagram: string;
  sequenceDiagram: string;
  deploymentDiagram: string;
}

export interface Blueprint {
  promptAnalysis: PromptAnalysis;
  decisions: Decision[];
  prd: PrdStructure;
  overview: string;
  features: { user: string[]; admin: string[]; system: string[] };
  techStack: Record<string, string[]>;
  database: { tables: Table[]; relationships: string[] };
  apis: ApiEndpoint[];
  folderStructure: { backend: string[]; frontend: string[] };
  awsArchitecture: {
    frontendHosting: string;
    backendHosting: string;
    database: string;
    storage: string;
    authentication: string;
    cdn: string;
    loadBalancer: string;
    flow: string[];
  };
  dockerArchitecture: { containers: string[]; flow: string[] };
  timeline: TimelinePhase[];
  security: string[];
  scalability: string[];
  futureEnhancements: string[];
  monitoring: Monitoring;
  estimatedCost: EstimatedCost;
  aiRecommendations: AiRecommendations;
  mermaid: Mermaid;
}

File: backend/app/schemas/blueprint.py
from typing import Any, Optional
from pydantic import BaseModel, Field


class BlueprintConfig(BaseModel):
    architectureStyle: str = Field(default="Monolithic")
    database: str = Field(default="PostgreSQL")
    backend: str = Field(default="FastAPI")
    frontend: str = Field(default="React")
    cloudProvider: str = Field(default="AWS")
    projectSize: str = Field(default="MVP")


class BlueprintRequest(BaseModel):
    idea: str
    config: BlueprintConfig


class PromptAnalysis(BaseModel):
    industry: str
    businessType: str
    complexity: str
    expectedUsers: str
    scale: str
    budget: str
    cloudRequirements: str
    compliance: str
    estimatedTimeline: str


class PrdDocumentMetadata(BaseModel):
    ownership: str
    deploymentTarget: str
    versionStatus: str


class PrdPersona(BaseModel):
    persona: str
    description: str


class PrdRequirement(BaseModel):
    requirement: str
    priority: str


class PrdUxDesign(BaseModel):
    interfaceOverview: str
    layoutDescription: str


class PrdNonFunctionalRequirement(BaseModel):
    requirement: str
    type: str


class PrdMetric(BaseModel):
    metric: str
    target: str


class PrdRisk(BaseModel):
    risk: str
    mitigation: str


class PrdStructure(BaseModel):
    documentMetadata: PrdDocumentMetadata
    executiveSummary: str
    userPersonas: list[PrdPersona]
    functionalRequirements: list[PrdRequirement]
    uxDesign: PrdUxDesign
    nonFunctionalRequirements: list[PrdNonFunctionalRequirement]
    metricsSuccess: list[PrdMetric]
    risksDependencies: list[PrdRisk]


# The AI is asked to return an object matching this shape. We keep it loose
# (dict-friendly) since the LLM output structure is fairly deep/nested and we
# don't want brittle validation to reject good-enough output.
class BlueprintResponse(BaseModel):
    promptAnalysis: PromptAnalysis
    prd: PrdStructure
    overview: str
    features: dict[str, list[str]]
    techStack: dict[str, Any]
    database: dict[str, Any]
    apis: list[dict[str, Any]]
    folderStructure: dict[str, Any]
    awsArchitecture: dict[str, Any]
    dockerArchitecture: dict[str, Any]
    timeline: list[dict[str, Any]]
    security: list[str]
    scalability: list[str]
    futureEnhancements: list[str]


class SqlGenerateRequest(BaseModel):
    database: dict[str, Any]
    dialect: str = Field(default="PostgreSQL")


class EndpointGenerateRequest(BaseModel):
    endpoint: dict[str, Any]
    framework: str = Field(default="FastAPI")


class CodeResponse(BaseModel):
    code: str
    language: str
